File: packages/server-go/internal/auth/capabilities.go
// Package auth — capabilities.go: AP-1 立场 ③ capability const 白名单
// (≤30, byte-identical 跟 spec §1 ③ + 蓝图 auth-permissions.md §1).
//
// 单源协议: 所有 endpoint authz 必须用本文件 const, 严禁 hardcode 字面
// permission name. 反约束 grep 锁:
//
//   git grep -nE 'HasCapability\("[a-z_]+"' packages/server-go/internal/api/
//   # 期望 0 hit (应改为 HasCapability(ctx, auth.CommitArtifact, scope))
//
// Spec锚: docs/implementation/modules/ap-1-spec.md §1 立场 ③ + §2 反约束 #1.
// 蓝图锚: docs/blueprint/auth-permissions.md §1 (ABAC + UI bundle 混合).
//
// admin god-mode capability 不在此白名单 — admin 走 /admin-api/* 单独
// middleware (admin.RequireAdmin), ADM-0 §1.3 红线 + spec §1 立场 ③ 字面.
package auth

// v1 capability 字面白名单 (spec §1 立场 ③ byte-identical).
//
// 改 = 改三处+: spec §1 ③ + 蓝图 auth-permissions.md §1 + acceptance
// `docs/qa/acceptance-templates/ap-1.md` §字面锁 + 此 const.
const (
	// channel scope (`*` / `channel:<id>`)
	ReadChannel   = "channel.read"
	WriteChannel  = "channel.write"
	DeleteChannel = "channel.delete"

	// artifact scope (`*` / `channel:<id>` / `artifact:<id>`)
	ReadArtifact     = "artifact.read"
	WriteArtifact    = "artifact.write"
	CommitArtifact   = "artifact.commit"
	IterateArtifact  = "artifact.iterate"
	RollbackArtifact = "artifact.rollback"

	// messaging
	MentionUser = "user.mention"
	ReadDM      = "dm.read"
	SendDM      = "dm.send"

	// channel admin (channel-scoped)
	ManageMembers = "channel.manage_members"
	InviteUser    = "channel.invite"
	ChangeRole    = "channel.change_role"
)

// ALL is the canonical ordered slice of capability literals — single
// source of truth for AP-4-enum (spec §0 立场 ①). Order is byte-identical
// to const block above (channel scope → artifact scope → messaging →
// channel admin); change one place to add a capability — init() rebuilds
// Capabilities map automatically; reflect-lint test守 ALL ↔ const drift.
//
// 反约束: 不准直接 mutate `Capabilities` map (反向 grep
// `Capabilities\[".*"\]\s*=` packages/server-go/internal/auth/ 仅 init() 1 hit).
var ALL = []string{
	ReadChannel,
	WriteChannel,
	DeleteChannel,
	ReadArtifact,
	WriteArtifact,
	CommitArtifact,
	IterateArtifact,
	RollbackArtifact,
	MentionUser,
	ReadDM,
	SendDM,
	ManageMembers,
	InviteUser,
	ChangeRole,
}

// Capabilities is the canonical full list (membership lookup + future
// CI lint reflection). AP-4-enum 立场 ① — auto-built from ALL via init();
// no direct literal init. Adding a new capability = add to ALL only.
var Capabilities map[string]bool

func init() {
	Capabilities = make(map[string]bool, len(ALL))
	for _, c := range ALL {
		Capabilities[c] = true
	}
}

// IsValidCapability is the single-source helper for handler-side
// capability validity checks (spec §0 立场 ③ — reasons.IsValid #496 同模式).
// Handlers must call this; direct `auth.Capabilities[name]` access in
// `internal/api/` is reverse-grep banned (count==0).
func IsValidCapability(name string) bool {
	return Capabilities[name]
}

File: packages/client/src/lib/capabilities.ts
// AP-2 client — capability label SSOT (i18n-ready, currently zh-CN literals).
//
// 立场承袭 (ap-2-spec.md §0.2 + content-lock + capability-dot-spec.md):
//   - 14 capability const byte-identical 跟 server `internal/auth/capabilities.go::ALL`
//     (改 = 改两处: server const + 此 LABEL_MAP)
//   - capabilityLabel(token) 单源 helper, 反 inline 字面散落 (反向 grep
//     `function capabilityLabel|export.*capabilityLabel` ==1 hit)
//   - 反 RBAC 角色名双语 (英 a/e/v/o + 中文 3 词) 0 hit (反 role bleed)
//
// CAPABILITY-DOT (post-rename): tokens are dot-notation `<domain>.<verb>` per
// 蓝图 auth-permissions.md §1 字面.
//
// 改 = 改三处: server `internal/auth/capabilities.go::ALL` + 此 LABEL_MAP +
// content-lock §1+§2 字面锁.

/** 14 capability token byte-identical 跟 server `auth.ALL` 顺序 */
export const CAPABILITY_TOKENS = [
  // channel scope
  'channel.read',
  'channel.write',
  'channel.delete',
  // artifact scope
  'artifact.read',
  'artifact.write',
  'artifact.commit',
  'artifact.iterate',
  'artifact.rollback',
  // messaging
  'user.mention',
  'dm.read',
  'dm.send',
  // channel acl
  'channel.manage_members',
  'channel.invite',
  'channel.change_role',
] as const;

export type CapabilityToken = (typeof CAPABILITY_TOKENS)[number];

/** 14 capability token → 中文 label SSOT (反 role 名双语漂). */
const LABEL_MAP: Record<CapabilityToken, string> = {
  'channel.read': '查看频道',
  'channel.write': '在频道发消息',
  'channel.delete': '删除频道',
  'artifact.read': '查看产物',
  'artifact.write': '编辑产物',
  'artifact.commit': '提交产物',
  'artifact.iterate': '迭代产物',
  'artifact.rollback': '回滚产物',
  'user.mention': '提及用户',
  'dm.read': '查看私信',
  'dm.send': '发送私信',
  'channel.manage_members': '管理频道成员',
  'channel.invite': '邀请用户',
  'channel.change_role': '调整成员能力',
};

/**
 * capabilityLabel — single-source helper to render a capability token
 * label. Unknown tokens render the raw token (forward-compat).
 *
 * 反约束: 调用方禁止 inline 字面 (如 `'查看频道'`); 走此 helper SSOT
 * (反向 grep `function capabilityLabel|export.*capabilityLabel` ==1 hit).
 */
export function capabilityLabel(token: string): string {
  if (token in LABEL_MAP) {
    return LABEL_MAP[token as CapabilityToken];
  }
  return token;
}

/** isKnownCapability — 反向断言 helper (跟 server `auth.IsValidCapability` 同精神). */
export function isKnownCapability(token: string): token is CapabilityToken {
  return token in LABEL_MAP;
}
